### ISSUE TITLE
Refactor ValidatorRegistration Pre-Consensus Process to Include Fee Recipient

### ISSUE THREAD

**Issue by olegshmuelov** [ssvlabs/ssv-spec#504]
The `ValidatorRegistration` duty currently does not include the fee recipient, but instead relies on a dynamic duty struct for all duties. The fee recipient is taken from a state that can be modified by an event handler (`FeeRecipientUpdate`) during the pre-consensus process. This introduces the potential for a data race, which may result in the signing root of the `ValidatorRegistration` struct being altered. If this happens, the reconstruction of the partial signatures will fail, causing the duty itself to fail.

A refactor is needed to ensure that the fee recipient is handled safely within the `ValidatorRegistration` process, preventing race conditions and ensuring the consistency of the duty.

**Comment by olegshmuelov** [comment]
@GalRogozinski @MatheusFranco99

**Comment by iurii-ssv** [comment]
@olegshmuelov I've looked into the relevant code and I want to clarify various scenarios that might happen (pls correct me if I'm wrong),

`ValidatorRegistration` object being signed [**always** uses `FeeRecipientAddress` of the corresponding `Runner`'s Share](https://github.com/ssvlabs/ssv/blob/419da26dc3f16999e5824f8430454392dc5df7a3/protocol/v2/ssv/runner/validator_registration.go#L210-L227) - the problem is once `Runner` has been initialized with some `Share` corresponding `FeeRecipientAddress` on that share **never changes** (until SSV node restarts - and once it does the latest version of Share is loaded from DB and supplied to `Runner`(s))

this means all operators sign the same `ValidatorRegistration` object every time (unless fee recipient changed for some operator during SSV node restart - meaning this operator will have a new fee recipient while other operators would still run with the older fee recipient until their respective SSV node restarts happen)

so there are actually 2 different manifestations of this problem that can happen:
- like you mention, validator registration duty will start failing if some operators switched to new fee recipient (due to SSV node restart) while others are still running with old fee recipient
- validator registration duty will not be aware of fee recipient update until cluster quorum of SSV nodes restart (after fee recipient update event comes through) - meaning SSV cluster will be sending validator registrations with old fee recipient for a while

and so the question is: "what can we do to make sure all SSV nodes in the cluster switch to new fee recipient ALL at the same time (same slot number I guess), and ideally do it before broadcasting corresponding validator registration"

- we can't just attach `FeeRecipientAddress` to PreConsensus message and expect other operators to accept it without verifying against their own source of data (because if they do they can be tricked into signing malicious validator registration)
- also note, we SubmitProposalPreparation periodically (which, as I understand, is an additional mechanism to relay info about up-to-date fee-recipient to Ethereum network)

I think the best we can do is to fetch the latest `FeeRecipientAddress` (either from `controller.validators` or from `Recipients interface`) in validator-registration duty instead of using `BaseRunner.Share`, thoughts ?

**Comment by iurii-ssv** [comment]
> In https://github.com/ssvlabs/ssv-spec/issues/368#issuecomment-2009613235, it was suggested to make independent structs for each duty type, which could be a more robust approach for preventing such issues.

Btw ^ that would be nice to do eventually, I don't think it's super relevant to this particular  issue though

**Comment by iurii-ssv** [comment]
> the runner holds a pointer to a copy of the share, meaning any updates (like FeeRecipientAddress) won’t be reflected after init. Passing the full *SSVShare ensures runners stay in sync with updates like fee recipient changes.

Right, it could be yet another option - but I don't see it being super viable because we'd need to synchronize access to `SSVShare` fields across go-routines because different go-routines update/read `FeeRecipientAddress` (and potentially other fields)

and we'd need to ensure it's synchronized properly everywhere `SSVShare` is used, so it's probably simpler to:
- keep synchronization to 1 place (`Recipients interface` in this case)
- and treat `SSVShare` as read-only snapshot everywhere in our code

**Comment by olegshmuelov** [comment]
I totally agree, passing `*SSVShare` could maybe work as a quick hotfix, but long-term it’s definitely cleaner to treat it as immutable and centralize all reads via `ValidatorStore`.

**Comment by olegshmuelov** [comment]
> `ValidatorRegistration` object being signed [**always** uses `FeeRecipientAddress` of the corresponding `Runner`'s Share](https://github.com/ssvlabs/ssv/blob/419da26dc3f16999e5824f8430454392dc5df7a3/protocol/v2/ssv/runner/validator_registration.go#L210-L227) - the problem is once `Runner` has been initialized with some `Share` corresponding `FeeRecipientAddress` on that share **never changes** (until SSV node restarts - and once it does the latest version of Share is loaded from DB and supplied to `Runner`(s))

I double checked and there's actually no copy being made in that line. The runner’s `BaseRunner.Share` points directly to `&options.SSVShare.Share`, so if `SSVShare.Share` is updated in-place, the change is reflected inside the runner too.

**Comment by olegshmuelov** [comment]
After revisiting the logic, I believe we can close this.

There’s technically a race condition where `FeeRecipientAddress` could change mid pre-consensus (since runners hold a mutable reference to the share), which may cause validator registration duties to fail temporarily due to signing root mismatch.

However, this failure is benign: the duty will retry in the next cycle, and once all nodes converge on the updated share, the signatures will match.

So while it's worth being aware of, it's not a critical issue. We might consider a cleanup/refactor in the future to make the `Share` immutable at runtime, but for now this doesn’t need to stay open.

**Comment by iurii-ssv** [comment]
> I double checked and there's actually no copy being made in that line. The runner’s `BaseRunner.Share` points directly to `&options.SSVShare.Share`, so if `SSVShare.Share` is updated in-place, the change is reflected inside the runner too.

Right, I see that both `BaseRunner.Share` and `Validator.Share` are pointing to the very same `SSVShare` instance (although it's not super obvious and relying on assumption isn't ideal)

but the problem with it is that there is no explicit synchronization between go-routine that [updates `FeeRecipientAddress`](https://github.com/ssvlabs/ssv/blob/9a68831664e44360865b268528ede0292e0ba1ec/operator/validator/task_executor.go#L84-L84) on `SSVShare` instance and another go-routine that's [reading `FeeRecipientAddress` in `BaseRunner`](https://github.com/ssvlabs/ssv/blob/9a68831664e44360865b268528ede0292e0ba1ec/protocol/v2/ssv/runner/validator_registration.go#L222-L222) - meaning it's a data-race from go memory model perspective (and hence to work correctly it needs additional explicit synchronization like I've described in https://github.com/ssvlabs/ssv-spec/issues/504#issuecomment-2785831012)

and that is essentially the reason why I suggest to treat `SSVShare` as "immutable snapshot", while propagating any updates to it through some sort of storage layer - eg. `validatorStore` (that will give us thread-safety)

**Comment by iurii-ssv** [comment]
> that still doesn’t solve the cross-operator timing issue e.g., some nodes might receive a fee recipient update mid-duty and sign different payloads. That’s a broader coordination challenge we might want to keep in mind.

True there will still be some timing discrepancy between operators receiving their respective events - I doubt we can do much about that, but also it's not that big of a deal either (since eventually all operators will receive the event, upgrade to the latest fee recipient and produce successful duty execution)

> I'm fully aligned on treating `SSVShare` as immutable and using `validatorStore` for dynamic reads [#504 (comment)](https://github.com/ssvlabs/ssv-spec/issues/504#issuecomment-2785965109)

I'll see if that's something we can do then
